### PR TITLE
Refactor tally objects in election-manager to be dictionaries and hold metadata with tally data

### PR DESCRIPTION
**Pull request by carolinemodic**
Refactors a few things about how we're storing tally information in election manager. In general previously we were using a lot of arrays instead of dictionaries to store things, this meant we had to find() within the array when building the tallies, instead of just using a dictionary and being able to access the right element. Additionally the metadata (undervotes, overvotes, number of ballots) were stored as completely different objects than the contest tallies and calculated completely separately. I think it makes more sense to calculate this all at the same time and store it in a combined structure. 

We were relying on the arrays in these objects to be the order of contests, and the order of candidates within contests, that we want to show on the tally reports. I don't think we should be relying on this for the ordering and should look back at the election definition to get the proper order. 

This is broken into two commits, commit 1 adds transforms back to the old format in votecounting.test.ts, in order to make sure that none of the actual tallies in the snapshots changed, the votecounting snapshot did not get updated in this commit, and passes successfully. Commit 2 removes these transforms and updates the snapshot to the new format. 

Note that the old getContestTallyMeta function is not yet removed, its now only used in the special overvote pairs screen which I don't fully understand yet so I will remove it and update that code separately.

### CODE REVIEW

File: apps/election-manager/src/__snapshots__/App.test.tsx.snap [hunk 1]
```diff
@@ -935,7 +935,7 @@ exports[`printing ballots, print report, and test decks 2`] = `
             <span
               class="sc-dQppl kMSBOR"
             >
-              0 ballots
+              39 ballots
```
Comment (carolinemodic) [inline apps/election-manager/src/__snapshots__/App.test.tsx.snap:938]: The previous code erroneously did not have a number of ballots for either/neither elections in the test deck tally report

File: apps/election-manager/src/__snapshots__/App.test.tsx.snap [hunk 2]
```diff
@@ -1826,7 +1826,7 @@ exports[`tabulating CVRs 1`] = `
             <td
               class="sc-fFubgz jVzFBN"
             >
-              54
+              53
```
Comment (carolinemodic) [inline apps/election-manager/src/__snapshots__/App.test.tsx.snap:1829]: The previous code erroneously counted votes for either/neither contests where the CVR is malformed and only contains one of the two contests, malformed CVRs like that will now be filtered out, causing the number here to decrease by 1 as there is 1 malformed CVR in the test set that has this contest but not the other.

File: apps/election-manager/src/__snapshots__/App.test.tsx.snap [hunk 3]
```diff
@@ -1880,7 +1880,7 @@ exports[`tabulating CVRs 1`] = `
             <td
               class="sc-fFubgz jVzFBN"
             >
-              50
+              49
```
Comment (carolinemodic) [inline apps/election-manager/src/__snapshots__/App.test.tsx.snap:1883]: The previous code erroneously counted votes for either/neither contests where the CVR is malformed and only contains one of the two contests, malformed CVRs like that will now be filtered out, causing the number here to decrease by 1 as there is 1 malformed CVR in the test set that has this contest but not the other.

File: apps/election-manager/src/utils/election.ts [hunk 4]
```diff
@@ -1,19 +1,50 @@
...
+    if ((contest as CandidateContest).allowWriteIns) {
+      return options.concat(writeInCandidate)
+    }
+    return options
+  }
+  if (contest.type === 'yesno') {
+    return [['yes'] as YesNoOption, ['no'] as YesNoOption]
+  }
+  throw new Error(`Unexpected contest type: ${contest.type}`)
+}
```
Comment (eventualbuddha) [inline apps/election-manager/src/utils/election.ts:46]: You can avoid the casting in this function by using `AnyContest`:

```ts
export function getContestOptionsForContest(
  contest: AnyContest
): readonly ContestOption[] {
  if (contest.type === 'candidate') {
    const options = contest.candidates
    if (contest.allowWriteIns) {
      return options.concat(writeInCandidate)
    }
    return options
  }
  if (contest.type === 'yesno') {
    return [['yes'], ['no']]
  }
  throw new Error(`Unexpected contest type: ${contest.type}`)
}
```

File: apps/election-manager/src/AppRoot.tsx [hunk 5]
```diff
@@ -84,10 +85,9 @@ const AppRoot: React.FC<Props> = ({ storage }) => {
   const [usbStatus, setUsbStatus] = useState(UsbDriveStatus.absent)
   const [recentlyEjected, setRecentlyEjected] = useState(false)
 
-  const [
-    fullElectionTally,
-    setFullElectionTally,
-  ] = useState<OptionalFullElectionTally>()
+  const [fullElectionTally, setFullElectionTally] = useState<FullElectionTally>(
```
Comment (eventualbuddha) [inline apps/election-manager/src/AppRoot.tsx:88]: Unnecessary explicit type parameter `<FullElectionTally>`, it'll be inferred.

**Review by eventualbuddha — APPROVED** [review]

**Pull request merged**